### ISSUE TITLE
aws-rds: Lambda invocation role support

### ISSUE THREAD

**Issue by underbluewaters** [aws/aws-cdk#14751]
RDS recently added support for [invoking lambda functions from within a postgres instance](https://aws.amazon.com/about-aws/whats-new/2021/04/amazon-rds-postgresql-integrates-aws-lambda). There's a certain amount of boilerplate necessary to set this up, including [assigning a role to the database instance](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/PostgreSQL-Lambda.html#PostgreSQL-Lambda-access) to control which lambda functions can be invoked. I don't believe there is any way to set this up currently using CDK. There's analogous support for setting up S3 import/export roles, as well as monitoring roles.

### Use Case
Currently all my database setup is included in my CDK app and I would like to keep it that way. Without support in CDK I have to manually run `aws rds add-role-to-db-instance` using the aws cli.

### Proposed Solution
Either support a flexible `initialPolicy` prop on RDS instances where an inline policy can be defined, or a named prop like `lambdaInvocationPolicy` with more focused support (though I'm not sure what the API would look like).

### Other
If there's an interim option for setting this up using the "layer 1" type api's I would love to hear about it.

**Comment by skinny85** [comment]
Hey @underbluewaters,

thanks for opening the issue. You're right that this will need some work to be implemented, similar to https://github.com/aws/aws-cdk/pull/10370. [Pull Requests are very much welcome](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md), as always 🙂.

If you want to use the L1 layer to set this, you can get the underlying `CfnDBInstance` from an L2 `Instance` using the `node.defaultChild` method, and then use [Escape hatches](https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html) to set any properties you need:

```ts
    const instance = new rds.DatabaseInstance(stack, 'Database', {
      engine: rds.DatabaseInstanceEngine.postgres({ version: rds.PostgresEngineVersion.VER_12_3 }),
      vpc,
    });
    const cfnInstance = instance.node.defaultChild as rds.CfnDBInstance;
    // change cfnInstance as you need...
```

Thanks,
Adam

**Comment by underbluewaters** [comment]
Thank you for the tip @skinny85. I was able to assign the necessary role using the following.

```typescript
    const cfnInstance = instance.node.defaultChild as rds.CfnDBInstance;

    const role = new iam.Role(this, "InvokeWorkerLambdas", {
      assumedBy: new iam.ServicePrincipal("rds.amazonaws.com"),
    });

    role.addToPolicy(
      new iam.PolicyStatement({
        actions: ["lambda:InvokeFunction"],
        effect: iam.Effect.ALLOW,
        resources: [
          `arn:aws:lambda:${props!.env!.region!}:${props!.env!
            .account!}:function:SendProjectInvites`,
        ],
      })
    );

    cfnInstance.associatedRoles = [
      { featureName: "Lambda", roleArn: role.roleArn },
    ];

```

**Comment by C-Deck** [comment]
Gave an attempt adding this. Hope it will help.

**Comment by rjmackay** [comment]
@TheRealAmazonKendra you commented on the previous PR for this asking to move away from `lambdaRole` style props. Is that a hard requirement for getting this implemented? Or would you be ok with just adding a new `lambdaRole` prop and doing that refactor as a separate task?

**Comment by kumsmrit** [comment]
Hi @underbluewaters

Thank you for raising this issue.  

While this gets picked up, you can unblock yourself today without waiting for the L2 change — the L1 `CfnDBInstance` already supports this, so you can set it using the `@aws-cdk/cfn-property-mixins` [CfnDBInstancePropsMixin](https://docs.aws.amazon.com/cdk/api/v2/docs/@aws-cdk_cfn-property-mixins.aws_rds.CfnDBInstancePropsMixin.html):

*Array merge note*: `associatedRoles` is a list, so `ArrayMergeStrategy.append()` can be applied here. Append will preserve any roles the L2 populates internally. More details about different array merge strategies: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.ArrayMergeStrategy.html

```ts
import { aws_rds as rdsMixin } from '@aws-cdk/cfn-property-mixins';
import { ArrayMergeStrategy, PropertyMergeStrategy } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';

// Build the role RDS assumes to invoke Lambda
const rdsLambdaRole = new iam.Role(this, 'RdsLambdaRole', {
  assumedBy: new iam.ServicePrincipal('rds.amazonaws.com'),
});
myFunction.grantInvoke(rdsLambdaRole);

// Attach it to the DB instance via the CfnProps mixin:
dbInstance.with(new rdsMixin.CfnDBInstancePropsMixin({
  associatedRoles: [{
    featureName: 'Lambda',
    roleArn: rdsLambdaRole.roleArn,
  }],
}, {
  strategy: PropertyMergeStrategy.combine({ arrays: ArrayMergeStrategy.append() }),
}));
```

This should unblock you immediately; let us know if this works for you.
